internal/app/client/client.go

```diff
@@ -0,0 +1,144 @@
+	"connectrpc.com/connect"
+	"golang.org/x/net/http2"
+)
+
+// Wrapper defines a wrapper around a client which can invoke requests based on a ClientCompatRequest.
+type Wrapper interface {
+	// Invoke invokes a request according to a given config
+	Invoke(context.Context, *v1alpha1.ClientCompatRequest) (*v1alpha1.ClientCompatResponse, error)
+}
+
+// Run runs the server according to server config read from the 'in' reader.
+func Run(ctx context.Context, _ []string, inReader io.ReadCloser, outWriter io.WriteCloser) error {
+	json := flag.Bool("json", false, "whether to use the JSON format for marshaling / unmarshaling messages")
+
+	flag.Parse()
+
+	// Read the server config from  the in reader
+	data, err := io.ReadAll(inReader)
+	if err != nil {
+		return err
+	}
+
+	codec := app.NewCodec(*json)
+
+	req := &v1alpha1.ClientCompatRequest{}
+	if err := codec.Unmarshal(data, req); err != nil {
+		return err
+	}
+
+	resp, err := invoke(ctx, req)
+	if err != nil {
+		return err
+	}
+	bytes, err := codec.Marshal(resp)
+	if err != nil {
+		return err
+	}
+	if _, err := outWriter.Write(bytes); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func invoke(ctx context.Context, req *v1alpha1.ClientCompatRequest) (*v1alpha1.ClientCompatResponse, error) {
+	var scheme string
+	if req.ServerTlsCert != nil {
+		scheme = "https://"
+	} else {
+		scheme = "http://"
+	}
+	urlString := scheme + net.JoinHostPort(req.Host, strconv.FormatUint(uint64(req.Port), 10))
+	serverURL, err := url.ParseRequestURI(urlString)
+	if err != nil {
+		return nil, errors.New("invalid url: %s" + urlString)
+	}
+
+	var transport http.RoundTripper
+	switch req.HttpVersion {
+	case v1alpha1.HTTPVersion_HTTP_VERSION_1:
+		transport = &http.Transport{}
+	case v1alpha1.HTTPVersion_HTTP_VERSION_2:
+		transport = &http2.Transport{
+			AllowHTTP: true,
+			DialTLSContext: func(ctx context.Context, network, addr string, cfg *tls.Config) (net.Conn, error) {
+				return net.Dial(network, addr)
+			},
+		}
+	case v1alpha1.HTTPVersion_HTTP_VERSION_3:
+		return nil, errors.New("HTTP/3 is not yet supported")
+	case v1alpha1.HTTPVersion_HTTP_VERSION_UNSPECIFIED:
+		return nil, errors.New("an HTTP version must be specified")
+	}
+
+	// Create client options based on protocol of the implementation
+	clientOptions := []connect.ClientOption{connect.WithHTTPGet()}
+	switch req.Protocol {
+	case v1alpha1.Protocol_PROTOCOL_GRPC:
+		clientOptions = append(clientOptions, connect.WithGRPC())
+	case v1alpha1.Protocol_PROTOCOL_GRPC_WEB:
+		clientOptions = append(clientOptions, connect.WithGRPCWeb())
+	case v1alpha1.Protocol_PROTOCOL_CONNECT:
+		// Do nothing
+	case v1alpha1.Protocol_PROTOCOL_UNSPECIFIED:
+		return nil, errors.New("a protocol must be specified")
+	}
+
+	if req.Codec == v1alpha1.Codec_CODEC_JSON {
+		clientOptions = append(clientOptions, connect.WithProtoJSON())
+	}
+
+	// TODO - How do we configure each compression algo? i.e.
```

Comment:
A TODO for this is fine.

But to answer the question: the reference client and server should support all of them, in case clients or servers under test support them. FWIW, the list of algorithms for the enum came from the actual protocol specs. Even though connect-go only supports "gzip" out-of-the-box, the Connect protocol spec also includes "br" and "zstd" (see _Content-Coding_ [here](https://connectrpc.com/docs/protocol/#unary-request)). And the gRPC protocol spec indicates "deflate" and "snappy" (see _Content-Coding_ [here](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests)).

In the future, there are implementations of all of these for Go:
* **br**: https://pkg.go.dev/github.com/andybalholm/brotli
* **zstd**: https://pkg.go.dev/github.com/klauspost/compress/zstd
* **deflate**: https://pkg.go.dev/compress/flate (in the Go stdlin actually)
* **snappy**: https://pkg.go.dev/github.com/golang/snappy

And they all have fairly similar APIs, similar to the decompressor/compressor reader/writer APIs of the gzip package (which is what connect-go includes by default).

Reply:
Gotcha. Thank you. Reworded the TODO and will add support in a follow-up